pallets/claim/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+                        // Subtract the claimed token from the TotalClaimable
+                        TotalClaimable::<T>::mutate(|required_amount| {
+                            *required_amount = required_amount.saturating_sub(*amount)
+                        });
+                        log::trace!("Claimed {amount:?} for {beneficiary:?}");
+                        Self::deposit_event(Event::<T>::Claimed {
+                            beneficiary,
+                            amount: *amount,
+                            payment_id,
+                        });
+                        None
+                    }
+                    // Account is not eligible to receive funds
+                    None => Err(Error::<T>::NotEligible)?,
+                };
+                Ok::<_, DispatchError>(())
+            })?;
+            Ok(())
+        }
+
+        fn do_add_beneficiaries(
+            beneficiaries: BTreeMap<T::AccountId, BalanceOf<T>>,
+        ) -> DispatchResult {
+            // Check that the pot has enough funds to cover for all the beneficiaries
+            let available_amount = Self::pot();
+            let mut required_amount = TotalClaimable::<T>::get();
+
+            beneficiaries
+                .iter()
+                .try_for_each::<_, DispatchResult>(|(account, amount)| {
+                    // Account already exists
+                    if let Some(old_amount) = Beneficiaries::<T>::get(account.clone()) {
+                        // We're giving the account less token compared to before
+                        if old_amount > *amount {
+                            // Subtract the difference from the required balance this pallet's account should have
+                            required_amount =
+                                required_amount.saturating_sub(old_amount.saturating_sub(*amount));
+                        } else {
+                            // We're giving the account more tokens compared to before
+                            // Add the difference to the required balance this pallet's account should have
+                            required_amount =
+                                required_amount.saturating_add(amount.saturating_sub(old_amount));
+                        }
+                        log::debug!("Beneficiary {account:?} already existing. Old amount: {old_amount:?} New amount: {amount:?}");
+                    } else {
+                        // Account doesn't exist. Add its token amount to the required amount this pallet's account should have
+                        required_amount = required_amount.saturating_add(*amount);
+                        log::trace!("Added beneficiary {account:?}. Can claim: {amount:?}");
+                    }
+
+                    // Cannot cover for all the tokens, raise an error
+                    if required_amount > available_amount {
+                        Err(Error::<T>::NotEnoughFunds)?;
+                    }
```

Comment:
If I'm understanding the logic correctly, I believe this check should be done out of the loop, as it may temporarily fail (while it shouldn't) during the iterations.

Also, I'm wondering whether it would make sense to make this entire function atomic: either it succeeds and `Beneficiaries` is updated, or it fails and the state is not changed at all.

Reply:
This is a good point indeed, but honestly I don't envision any of the add/remove beneficiaries being ever called after having added beneficiaries the first time. On top of this, considering they could only be called by a Super User or Treasury, we can safely assume that they won't be that frequent/superficial to cause this kind of issues.
I already felt overdesigning with these mechanisms, and thus I don't feel complicating the logic furthermore for something that will likely never happen.
But I'm happy to discuss further you think it's indeed worth the effort :)

Reply:
Extrinsic execution should already be atomic: if it fails in the middle of the execution, the state will stay untouched

Reply:
Right, the atomicity is guaranteed by the framework, makes sense. But I still believe the first point should be addressed: in the current state, the update might fail even if it is correct.
For example, imagine having a pot of `10`, distributed like `(addr1, 5), (addr2, 5)`; updating that to `(addr1, 6), (addr2, 4)` might fail with `NotEnoughFunds`, depending on the order of the beneficiaries map.

Reply:
Ah right, good catch !